src/common/src/util/value_encoding/row_encoding.rs

```diff
@@ -0,0 +1,221 @@
+    datum_num: Width,
+    encoded_datum_num: Vec<u8>,
+}
+
+impl Serializer {
+    pub fn new(column_ids: &[ColumnId]) -> Self {
+        let mut encoded_column_ids = Vec::with_capacity(column_ids.len() * 4);
+        for id in column_ids {
+            encoded_column_ids.put_i32_le(id.get_id());
+        }
+        let datum_num = match column_ids.len() {
+            n if n <= u8::MAX as usize => Width::Mid(n as u8),
+            n if n <= u16::MAX as usize => Width::Large(n as u16),
+            n if n <= u32::MAX as usize => Width::Extra(n as u32),
+            _ => unreachable!("the number of columns exceeds u32"),
+        };
+        let mut encoded_datum_num = vec![];
+        serialize_width(datum_num, &mut encoded_datum_num);
+        Self {
+            encoded_column_ids,
+            datum_num,
+            encoded_datum_num,
+        }
+    }
+
+    pub fn serialize_row_column_aware(&self, row: impl Row) -> Vec<u8> {
+        let mut encoding = RowEncoding::new();
+        encoding.encode(row.iter());
+        self.serialize(encoding)
+    }
+
+    fn serialize(&self, mut encoding: RowEncoding) -> Vec<u8> {
+        let mut row_bytes = vec![];
+        encoding.set_width(&self.datum_num);
+        row_bytes.put_u8(encoding.flag);
+        row_bytes.extend(self.encoded_datum_num.iter());
+        row_bytes.extend(self.encoded_column_ids.iter());
+        row_bytes.extend(encoding.offsets.iter());
+        row_bytes.extend(encoding.buf.iter());
+
+        row_bytes
+    }
+}
+
+pub struct Deserializer<'a> {
+    needed_column_ids: HashMap<i32, usize>,
+    schema: &'a [DataType],
+}
+
+impl<'a> Deserializer<'a> {
+    pub fn new(column_ids: &'a [ColumnId], schema: &'a [DataType]) -> Self {
+        Self {
+            needed_column_ids: column_ids
+                .iter()
+                .enumerate()
+                .map(|(i, c)| (c.get_id(), i))
+                .collect::<HashMap<_, _>>(),
+            schema,
+        }
+    }
+
+    pub fn decode(&self, mut encoded_bytes: &[u8]) -> Result<Vec<Datum>> {
+        let flag = encoded_bytes.get_u8();
+        let nums_bytes = match flag & 0b1100 {
+            0b0100 => 1,
+            0b1000 => 2,
+            0b1100 => 4,
+            _ => return Err(ValueEncodingError::InvalidFlag(flag)),
+        };
+        let offset_bytes = match flag & 0b11 {
+            0b01 => 1,
+            0b10 => 2,
+            0b11 => 4,
+            _ => return Err(ValueEncodingError::InvalidFlag(flag)),
+        };
+        let datum_num = deserialize_width(nums_bytes, &mut encoded_bytes);
+        let offsets_start_idx = 4 * datum_num;
+        let data_start_idx = offsets_start_idx + datum_num * offset_bytes;
+        let offsets = &encoded_bytes[offsets_start_idx..data_start_idx];
+        let data = &encoded_bytes[data_start_idx..];
+        let mut datums = vec![None; self.schema.len()];
+        for i in 0..datum_num {
+            let this_id = encoded_bytes.get_i32_le();
```

Comment:
So we're not using different lengths to encode the column ID according to the `Width`? Sounds okay for now as the type of `ColumnId` is signed, which is strange.

Actually I prefer totally ignore the `Width` and always use u32 for the `datum_num` for now. 😄 Compared to the bytes wasted by the column IDs, the `datum_num` will not be a big deal.

Reply:
Sure! `Width` is dropped now and could be introduced later if we want to optimize `ColumnId`'s representation